### ISSUE TITLE
txhashset_write: corrupted storage?

### ISSUE THREAD

**Issue by bladedoyle** [mimblewimble/grin#2871]
**Describe the bug**
Attempts to sync the blockchain (floonet) fail with: "Segmentation fault" following "Corrupted storage" error.  The initial problem appears to be:

```
20190602 04:59:21.026 DEBUG grin_chain::txhashset::txhashset - txhashset_replace: move from "/tmp" to "/server/.grin"
20190602 04:59:21.026 ERROR grin_chain::txhashset::txhashset - hashset_replace fail on txhashset. err: Invalid cross-device link (os error 18)
```

**To Reproduce**
Sync a floonet node from scratch.

**Relevant Information**
Node is running with Docker, on google cloud with /server filesystem mounted as cloud storage.  Full Logs:
```
20190602 04:59:20.876 DEBUG grin_chain::txhashset::txhashset - txhashset: rebuild_index: 158419 UTXOs, took 0s
20190602 04:59:20.912 DEBUG grin_chain::chain - txhashset_write: finished validating and rebuilding
20190602 04:59:20.912 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetValidation { kernels: 181140, kernel_total: 181153, rproofs: 158400, rproof_total: 230885 } -> TxHashsetSave
20190602 04:59:21.025 DEBUG grin_chain::chain - txhashset_write: finished committing the batch (head etc.)
20190602 04:59:21.026 DEBUG grin_chain::txhashset::txhashset - txhashset_replace: move from "/tmp" to "/server/.grin"
20190602 04:59:21.026 ERROR grin_chain::txhashset::txhashset - hashset_replace fail on txhashset. err: Invalid cross-device link (os error 18)
20190602 04:59:21.071 ERROR grin_servers::common::adapters - Failed to save txhashset archive: TxHashSetErr: txhashset replacing fail 
 Cause: Unknown 
 Backtrace: 
20190602 04:59:21.072 DEBUG grin_p2p::protocol - handle_payload: txhashset archive for 0310fa0bac16 at 155933, DONE. Data Ok: true
20190602 04:59:21.072 ERROR grin_servers::grin::sync::state_sync - state_sync: error = Chain(Error { inner: 

TxHashSetErr: txhashset replacing fail }). restart fast sync
20190602 04:59:21.128 DEBUG grin_servers::grin::sync::state_sync - state_sync: before txhashset request, header head: 158814 / 00ee00837bcd, txhashset_head: 155933 / 0310fa0bac16
20190602 04:59:21.128 DEBUG grin_p2p::peer - Asking 47.90.124.195:3414 for txhashset archive at 155933 0310fa0bac16.
20190602 04:59:21.129 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetSave -> TxHashsetDownload { start_time: 2019-06-02T04:59:21.129157467Z, prev_update_time: 2019-06-02T04:59:21.129157640Z, update_time: 2019-06-02T04:59:21.129157713Z, prev_downloaded_size: 0, downloaded_size: 0, total_size: 0 }
20190602 04:59:30.041 DEBUG grin_p2p::peer - Client 173.230.154.243:13414 connection lost: Connection(Custom { kind: ConnectionAborted, error: StringError("read_exact") })
20190602 04:59:30.344 DEBUG grin_servers::grin::seed - peer_connect: ignore a duplicate request to 173.230.154.243:13414. previous connecting time: 04:59:10.337
20190602 04:59:30.344 DEBUG grin_servers::grin::seed - peer_connect: ignore a duplicate request to 95.216.208.243:13414. previous connecting time: 04:59:10.337
20190602 04:59:30.345 DEBUG grin_servers::grin::seed - peer_connect: ignore a duplicate request to 35.231.240.252:13414. previous connecting time: 04:59:10.337
20190602 04:59:30.345 DEBUG grin_servers::grin::seed - monitor_peers: on 0.0.0.0:3414, 10 connected (5 most_work). all 272 = 13 healthy + 0 banned + 259 defunct
20190602 04:59:30.345 DEBUG grin_p2p::peers - clean_peers PeerAddr(V4(173.230.154.243:13414)), not connected
20190602 04:59:30.346 DEBUG grin_p2p::peers - clean_peers PeerAddr(V4(95.216.208.243:13414)), not connected
20190602 04:59:43.476 DEBUG grin_p2p::peer - Client 78.46.85.142:13414 connection lost: Connection(Custom { kind: ConnectionAborted, error: StringError("read_exact") })
20190602 04:59:45.638 DEBUG grin_p2p::peer - Client 94.130.64.25:13414 connection lost: Connection(Custom { kind: ConnectionAborted, error: StringError("read_exact") })
20190602 04:59:47.856 DEBUG grin_p2p::protocol - handle_payload: txhashset archive for 0310fa0bac16 at 155933. size=185734851
20190602 04:59:50.351 DEBUG grin_servers::grin::seed - peer_connect: ignore a duplicate request to 112.64.68.109:13414. previous connecting time: 04:59:30.344
20190602 04:59:50.351 DEBUG grin_servers::grin::seed - monitor_peers: on 0.0.0.0:3414, 8 connected (4 most_work). all 272 = 13 healthy + 0 banned + 259 defunct
20190602 04:59:50.351 DEBUG grin_p2p::peers - clean_peers PeerAddr(V4(78.46.85.142:13414)), not connected
20190602 04:59:50.352 DEBUG grin_p2p::peers - clean_peers PeerAddr(V4(94.130.64.25:13414)), not connected
20190602 04:59:50.373 DEBUG grin_p2p::peer - connect: handshaking with Ok(V4(173.230.154.243:13414))
20190602 04:59:50.394 DEBUG grin_p2p::handshake - Connected! Cumulative 16532716662 offered from PeerAddr(V4(173.230.154.243:13414)) "MW/Grin 1.1.0-beta.2" HEADER_HIST | TXHASHSET_HIST | PEER_LIST | TX_KERNEL_HASH | FULL_NODE
20190602 04:59:50.395 DEBUG grin_p2p::peers - Saving newly connected peer 173.230.154.243:13414.
20190602 04:59:50.395 DEBUG grin_p2p::store - save_peer: PeerAddr(V4(173.230.154.243:13414)) marked Healthy
20190602 04:59:50.533 DEBUG grin_p2p::peer - connect: handshaking with Ok(V4(95.216.208.243:13414))
20190602 04:59:50.678 DEBUG grin_p2p::peer - connect: handshaking with Ok(V4(106.12.13.235:13414))
20190602 04:59:50.714 DEBUG grin_p2p::handshake - Connected! Cumulative 16532716662 offered from PeerAddr(V4(95.216.208.243:13414)) "MW/Grin 1.0.3" HEADER_HIST | TXHASHSET_HIST | PEER_LIST | TX_KERNEL_HASH | FULL_NODE
20190602 04:59:50.715 DEBUG grin_p2p::peers - Saving newly connected peer 95.216.208.243:13414.
20190602 04:59:50.715 DEBUG grin_p2p::store - save_peer: PeerAddr(V4(95.216.208.243:13414)) marked Healthy
20190602 04:59:51.001 DEBUG grin_p2p::handshake - Connected! Cumulative 100000 offered from PeerAddr(V4(106.12.13.235:13414)) "MW/Grin 1.1.0-beta.2" HEADER_HIST | TXHASHSET_HIST | PEER_LIST | TX_KERNEL_HASH | FULL_NODE
20190602 04:59:51.002 DEBUG grin_p2p::peers - Saving newly connected peer 106.12.13.235:13414.
20190602 04:59:51.002 DEBUG grin_p2p::store - save_peer: PeerAddr(V4(106.12.13.235:13414)) marked Healthy
20190602 04:59:56.829 DEBUG grin_servers::common::adapters - locator: [1c64fc61e248, 0b45a0035ed3, 075b5c9800d6, 144206984364, 1a55475eb271, 11ad1ace06c9, 10c9328b576c, 27b17b01858b, 33029adfd378, 19ae6916bfeb, 21a4d9761a48, 12088136fc68, 013921871deb, 1ac5048490f0, 02539eafd7f5, 171dc236d082, 029b54270c62, edc758c1370d]
20190602 04:59:56.830 ERROR grin_store::types - Corrupted storage, could not read an entry from data file: IOErr("failed to fill whole buffer", UnexpectedEof)
20190602 05:00:04.815 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetDownload { start_time: 2019-06-02T04:59:47.857225295Z, prev_update_time: 2019-06-02T05:00:03.486042515Z, update_time: 2019-06-02T05:00:03.486088003Z, prev_downloaded_size: 185712000, downloaded_size: 185734851, total_size: 185734851 } -> TxHashsetSetup
20190602 05:00:04.815 DEBUG grin_chain::chain - txhashset_write: body_head - 0310fa0bac16, 155933, header_head - 00ee00837bcd, 158814, sync_head - 00ee00837bcd, 158814
20190602 05:00:06.831 ERROR grin_p2p::protocol - handle_payload: BanReason BanReason { ban_reason: FraudHeight }
20190602 05:00:06.842 DEBUG grin_p2p::peer - Client 106.12.13.235:13414 connection lost: Connection(Custom { kind: ConnectionAborted, error: StringError("read_exact") })
20190602 05:00:08.695 ERROR grin_store::types - Corrupted storage, could not read an entry from data file: IOErr("failed to fill whole buffer", UnexpectedEof)
20190602 05:00:08.695 ERROR grin_chain::chain - txhashset_write: corrupted storage? oldest_height is 0 when check_txhashset_needed. state sync is needed
20190602 05:00:08.696 DEBUG grin_chain::txhashset::txhashset - zip_write on path: "/tmp"
20190602 05:00:09.490 DEBUG grin_store::leaf_set - leaf_set: copying rewound file /tmp/txhashset/output/pmmr_leaf.bin.0310fa0bac16 to /tmp/txhashset/output/pmmr_leaf.bin
/run.sh: line 59:    18 Segmentation fault      (core dumped) grin ${NET_FLAG} server run
```

**Desktop (please complete the following information):**
grin v1.0.3

**Comment by bladedoyle** [comment]
Workaround:  To avoid use of txhashset zip files, I was able to use a blockchain copied from another (already synced) node.

Suggestion:  Use a temporary directory on the same filesystem as the blockchain for downloading txhashset archives.

**Comment by hashmap** [comment]
Duplicate of https://github.com/mimblewimble/grin/issues/2709 fixed in 1.1. I’m not sure it makes sense to backport it to 1.0.3, hf is coming.
